### PR TITLE
Bipart hpp

### PR DESCRIPTION
**Pull request by james-d-mitchell**
Not completely finished with this, but nearly.

### CODE REVIEW

**Comment by james-d-mitchell** [comment]
Finished and ready for review @Joseph-Edwards

File: include/libsemigroups/bipart.hpp [hunk 1]
```diff
@@ -24,29 +24,256 @@
...
+  //! \ingroup bipart_group
+  //!
+  //! \brief Namespace for Bipartition helper functions.
+  //!
+  //! This namespace contains helper functions for the Bipartition class.
+  namespace bipartition {
+    //! \brief Return the identity bipartition with the same degree as the
+    //! given bipartition.
+    //!
+    //! \param f the bipartition
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:109]:
```suggestion
    //! \param f the bipartition.
```

File: include/libsemigroups/bipart.hpp [hunk 2]
```diff
@@ -24,29 +24,256 @@
...
+    //! \brief Return the underlying partition of a Bipartition object.
+    //!
+    //! The *underlying partition* of a bipartition \c x is the
+    //! partition of a subset \f$P\f$ of \f$\{-n, \ldots, -1\}\cup \{1, \ldots,
+    //! n\}\f$ such that:
+    //! * \f$\{|x|\mid x\in P\} = \{1, \ldots, n\}\f$;
+    //! * a block of the partition consists of negative numbers if and only if
+    //! the corresponding block of \c x is a transverse block.
+    //!
+    //! \param x the Bipartition.
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:131]:
```suggestion
    //! \param x the bipartition.
```

File: include/libsemigroups/bipart.hpp [hunk 3]
```diff
@@ -24,29 +24,256 @@
...
+    //! \f$O(n)\f$ where \f$n\f$ is the degree().
+    [[nodiscard]] std::vector<std::vector<int32_t>>
+    underlying_partition(Bipartition const&);
+
+    //! \brief Validates a bipartition.
+    //!
+    //! This function validates a Bipartition object, and throws an exception if
+    //! the object is not valid.
+    //!
+    //! \param x the bipartition
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:149]:
```suggestion
    //! \param x the bipartition.
```

File: include/libsemigroups/bipart.hpp [hunk 4]
```diff
@@ -24,29 +24,256 @@
...
+      // checks for this argument type are done in validate
+    }
+  }  // namespace detail
+
+  //! \ingroup bipart_group
+  //!
+  //! \brief A Blocks object represents a signed partition of the set
   //! \f$\{0, \ldots, n - 1\}\f$.
   //!
+  //! Defined ``bipart.hpp``.
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:275]:
```suggestion
  //! Defined in ``bipart.hpp``.
```

File: include/libsemigroups/bipart.hpp [hunk 5]
```diff
@@ -62,33 +289,40 @@ namespace libsemigroups {
...
+    //! \brief Construct a blocks object from iterators.
     //!
     //! The degree of the blocks object constructed is `last - first / 2`.
     //!
-    //! \param first the index of the block containing the first point
-    //! \param last  the index of the block containing the first point
+    //! \param first iterator pointing to the index of the block containing the
+    //! first point
+    //! \param last iterator pointing one past the index of the block containing
+    //! the last point
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:325]:
```suggestion
    //! \param first iterator pointing to the index of the block containing the
    //! first point.
    //! \param last iterator pointing one past the index of the block containing
    //! the last point.
```

File: include/libsemigroups/bipart.hpp [hunk 6]
```diff
@@ -113,28 +347,57 @@ namespace libsemigroups {
     //! Linear in \p degree.
     explicit Blocks(size_t degree) : _blocks(degree), _lookup() {}
 
+    //! \brief Constructs a Blocks object from a vector of vectors of integers.
+    //!
+    //! This function constructs a Blocks object from a vector of vectors of
+    //! (signed) integers, so that the blocks consisting of negative values are
+    //! transverse and those consisting of positive values are not.
+    //!
+    //! \param blocks the blocks
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:356]:
```suggestion
    //! \param blocks the blocks.
```

File: include/libsemigroups/bipart.hpp [hunk 7]
```diff
@@ -113,28 +347,57 @@ namespace libsemigroups {
...
+    //!
+    //! \throws LibsemigroupsException if the set consisting of the absolute
+    //! values of the entries in \p blocks is not \f${1, \ldots, n\}\f$ where
+    //! \f$n\f$ is the maximum such value.
+    //! \throws LibsemigroupsException if \c 0 is an item in any block.
+    //! \throws LibsemigroupsException if any block is empty.
+    //! \throws LibsemigroupsException if any block contains both negative and
+    //! positive values..
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:364]:
```suggestion
    //! \throws LibsemigroupsException if any block contains both negative and
    //! positive values.
```

File: include/libsemigroups/bipart.hpp [hunk 8]
```diff
@@ -143,18 +406,80 @@ namespace libsemigroups {
...
+      return is_transverse_block_no_checks(i, val);
+    }
+
+    //! \brief Check if a block is a transverse block.
+    //!
+    //! This function returns \c true if the block with index \p index is a
+    //! transverse (or signed) block and it returns \c false if it is not
+    //! transverse (or unsigned).
+    //!
+    //! \param index the index of a block
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:439]:
```suggestion
    //! \param index the index of a block.
```

File: include/libsemigroups/bipart.hpp [hunk 9]
```diff
@@ -143,18 +406,80 @@ namespace libsemigroups {
...
+      return _lookup[index];
+    }
+
+    //! \brief Check if a block is a transverse block.
+    //!
+    //! This function returns \c true if the block with index \p index is a
+    //! transverse (or signed) block and it returns \c false if it is not
+    //! transverse (or unsigned).
+    //!
+    //! \param index the index of a block
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:462]:
```suggestion
    //! \param index the index of a block.
```

File: include/libsemigroups/bipart.hpp [hunk 10]
```diff
@@ -383,31 +705,60 @@ namespace libsemigroups {
...
+      // TODO(2) better error
+      return _blocks.at(i);
+    }
 
-  // Forward decl
-  class Bipartition;
+    //! \brief Validate the arguments, construct a Blocks object, and validate
+    //! it.
+    //!
+    //! \tparam Container the type of the parameter \p cont
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:728]:
```suggestion
    //! \tparam Container the type of the parameter \p cont.
```

File: include/libsemigroups/bipart.hpp [hunk 11]
```diff
@@ -383,31 +705,60 @@ namespace libsemigroups {
...
+    [[nodiscard]] static Blocks make(Container const& cont);
 
-  //! Validates a bipartition.
-  //!
-  //! \param x the bipartition
+    //! \copydoc make(Container const&)
+    [[nodiscard]] static Blocks
+    make(std::initializer_list<std::vector<int32_t>> const& cont) {
+      return make<std::initializer_list<std::vector<int32_t>>>(cont);
+    }
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:744]: Should this be called `to_blocks` like with presentations?

File: include/libsemigroups/bipart.hpp [hunk 12]
```diff
@@ -818,51 +1197,52 @@ namespace libsemigroups {
...
+    [[nodiscard]] size_t rank() const;
 
-    //! Returns the number of blocks in a Bipartition.
+    //! \brief Return the number of blocks in a Bipartition.
     //!
     //! This function returns the number of parts in the partition that
     //! instances of this class represent.
     //!
-    //! \returns The number of blocks in a Blocks object.
+    //! \returns The number of blocks..
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:1236]:
```suggestion
    //! \returns The number of blocks.
```

File: include/libsemigroups/bipart.hpp [hunk 13]
```diff
@@ -895,31 +1272,59 @@ namespace libsemigroups {
...
+    //!
+    //! \warning This function does no checks on its arguments.
+    [[nodiscard]] bool is_transverse_block_no_checks(size_t index) const;
+
+    //! \brief Check if a block is a transverse block.
+    //!
+    //! A block of a biparition is *transverse* if it contains integers less
+    //! than and greater than \f$n\f$, which is the degree of the bipartition.
+    //!
+    //! \param index the index of a block
```
Comment (Joseph-Edwards) [inline include/libsemigroups/bipart.hpp:1300]:
```suggestion
    //! \param index the index of a block.
```

File: include/libsemigroups/transf.hpp [hunk 14]
```diff
@@ -667,7 +668,7 @@ namespace libsemigroups {
     //!
     //! This function returns a newly constructed partial transformation with
     //! degree equal to the degree of \c N that fixes every value from \c 0
-    //! to \c N.
+    //! to \c N - 1.
```
Comment (Joseph-Edwards) [inline include/libsemigroups/transf.hpp:671]:
```suggestion
    //! to `N - 1`.
```

**Review by Joseph-Edwards — CHANGES_REQUESTED** [review]
Modulo the minor punctuation and potential naming changes I've left in the comments, this looks good to me!

File: include/libsemigroups/bipart.hpp (hunk 11 above)
Comment (james-d-mitchell) [reply include/libsemigroups/bipart.hpp:744]: Yeah you're almost certainly correct, I thought having `Blocks::make` was close enough but I see I am just being lazy :)

**Comment by james-d-mitchell** [comment]
Thanks @Joseph-Edwards for the thorough review!

**Comment by james-d-mitchell** [comment]
Ready to go as far as I'm concerned @Joseph-Edwards, thanks again!

**Review by Joseph-Edwards — APPROVED** [review]

**Pull request merged**